File: ios/Sources/ViewModel/ChatViewModel.swift
import Combine
import Foundation

@MainActor
final class ChatViewModel: ObservableObject {
    @Published var messages: [ChatMessage] = []
    @Published var streaming: String? = nil
    @Published var sources: [Source] = []
    @Published var palette: Palette? = nil   // 共感覚パレット（情調→伝統色）
    @Published var isLoading: Bool = false
    @Published var errorText: String? = nil

    private let api = BungoAPI(baseURL: Config.baseURL, apiKey: Config.apiKey)

    func send(_ text: String) {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, !isLoading else { return }

        let history = messages
        messages.append(ChatMessage(role: "user", content: trimmed))
        streaming = ""
        sources = []
        palette = nil
        errorText = nil
        isLoading = true

        Task {
            do {
                try await api.streamChat(
                    ChatRequest(message: trimmed, history: history, topK: 5)
                ) { [weak self] event in
                    guard let self else { return }
                    switch event {
                    case .token(let content):
                        self.streaming = (self.streaming ?? "") + content
                    case .sources(let newSources, let newPalette):
                        self.sources = newSources
                        self.palette = newPalette
                    case .done:
                        if let finalText = self.streaming {
                            self.messages.append(ChatMessage(role: "assistant", content: finalText))
                        }
                        self.streaming = nil
                        self.isLoading = false
                    case .error(let message):
                        self.errorText = message
                        self.streaming = nil
                        self.isLoading = false
                    }
                }
            } catch {
                self.errorText = "通信エラー: \(error.localizedDescription)"
                self.streaming = nil
                self.isLoading = false
            }
        }
    }
}

File: android/app/src/main/java/com/bungo/rag/ChatViewModel.kt
package com.bungo.rag

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.bungo.rag.data.BungoApi
import com.bungo.rag.data.ChatMessage
import com.bungo.rag.data.ChatRequest
import com.bungo.rag.data.Source
import com.bungo.rag.data.StreamEvent
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

data class ChatUiState(
    val messages: List<ChatMessage> = emptyList(),
    val streaming: String? = null,          // 生成中の暫定アシスタント応答
    val sources: List<Source> = emptyList(),
    val isLoading: Boolean = false,
    val error: String? = null,
)

class ChatViewModel : ViewModel() {

    private val api = BungoApi(BuildConfig.BASE_URL, BuildConfig.API_KEY)

    private val _ui = MutableStateFlow(ChatUiState())
    val ui: StateFlow<ChatUiState> = _ui.asStateFlow()

    fun send(text: String, topK: Int = 5) {
        val message = text.trim()
        if (message.isEmpty() || _ui.value.isLoading) return

        val history = _ui.value.messages
        _ui.update {
            it.copy(
                messages = it.messages + ChatMessage("user", message),
                streaming = "",
                sources = emptyList(),
                isLoading = true,
                error = null,
            )
        }

        viewModelScope.launch {
            val builder = StringBuilder()
            try {
                withContext(Dispatchers.IO) {
                    api.streamChat(
                        ChatRequest(message = message, history = history, topK = topK)
                    ) { event ->
                        when (event) {
                            is StreamEvent.Token -> {
                                builder.append(event.content)
                                _ui.update { it.copy(streaming = builder.toString()) }
                            }
                            is StreamEvent.Sources ->
                                _ui.update { it.copy(sources = event.sources) }
                            is StreamEvent.Error ->
                                _ui.update { it.copy(error = event.message) }
                            StreamEvent.Done -> Unit
                        }
                    }
                }
                commitAssistant(builder.toString())
            } catch (e: Exception) {
                _ui.update {
                    it.copy(
                        isLoading = false,
                        streaming = null,
                        error = "通信エラー: ${e.message ?: e.javaClass.simpleName}",
                    )
                }
            }
        }
    }

    private fun commitAssistant(full: String) {
        _ui.update {
            val finalText = full.ifBlank { it.error?.let { e -> "⚠️ $e" } ?: "" }
            it.copy(
                messages = if (finalText.isBlank()) it.messages
                else it.messages + ChatMessage("assistant", finalText),
                streaming = null,
                isLoading = false,
            )
        }
    }

    fun reset() {
        _ui.value = ChatUiState()
    }

    override fun onCleared() {
        api.close()
        super.onCleared()
    }
}
